### PR TITLE
feat(Table): add filter to colum settings

### PR DESCRIPTION
**Pull request by polikashina**
Added search for column settings for withTableSettings HOC
https://st.yandex-team.ru/UXRFC-485

### CODE REVIEW

File: src/components/Table/README.md [hunk 1]
```diff
@@ -251,12 +252,15 @@ const MyTable1 = withTableSettings({sortable: false})(Table);
...
+| :--------------------- | :----------------------------------------------------------- | :------------------------------------------------------: |
+| settingsPopupWidth     | TableColumnSetup pop-up width                                |                      `number` `fit`                      |
+| settings               | Current settings                                             |                   `TableSettingsData`                    |
+| updateSettings         | Settings update handle                                       |       `(data: TableSettingsData) => Promise<void>`       |
+| renderControls         | Allows to render custom actions                              |                     `RenderControls`                     |
+| filterPlaceholder      | Text that appears in the control when no search value is set |                         `string`                         |
+| filterEmptyPlaceholder | Text that appears when no one item is found                  |                         `string`                         |
+| filterItems            | Function for filtering items                                 | `(item: TableColumnSetupItem, value: string) => boolean` |
```
Comment (amje) [inline src/components/Table/README.md:263]: All the props should have `settings` in their names to show belonging to the HOC. For example:
`filterSettings`
`settingsFilterPlaceholder`
`settingsFilterEmptyMessage` (to be consistent with `emptyMessage` of the `Table` props)

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx [hunk 2]
```diff
@@ -242,6 +246,17 @@ const mapItemDataToProps = (item: TableColumnSetupItem): ListItemCommonProps =>
     };
 };
 
+const defaultFilterItemsFn = (item: TableColumnSetupItem, value: string) => {
+    return typeof item.title === 'string'
+        ? item.title.toLowerCase().includes(value.toLowerCase())
```
Comment (Raubzeug) [inline src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:251]: I would also trim value.

**Review by Raubzeug — CHANGES_REQUESTED** [review]

File: src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx [hunk 3]
```diff
@@ -242,6 +246,17 @@ const mapItemDataToProps = (item: TableColumnSetupItem): ListItemCommonProps =>
     };
 };
 
+const defaultFilterItemsFn = (item: TableColumnSetupItem, value: string) => {
```
Comment (IsaevAlexandr) [inline src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:249]: may be it's passible to use `useListFilter` from [this](https://github.com/gravity-ui/uikit/blob/main/src/components/TreeSelect/__stories__/components/WithFiltrationAndControlsExample.tsx#L48) example and not realise filtration logic from stretch?

it is also welcome for the development of a hook component for filtering

@polikashina @amje @Raubzeug  What do you think?

File: src/components/Table/README.md (hunk 1 above)
Comment (polikashina) [reply src/components/Table/README.md]: [fixed](https://github.com/gravity-ui/uikit/pull/1627/commits/49395e65e53f303aa73968265b49bf54cd6c4d35)

File: src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx (hunk 3 above)
Comment (polikashina) [reply src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx]: [fixed](https://github.com/gravity-ui/uikit/pull/1627/commits/05233de4c44a860d5dfe38e6318cc68bb87f607b)

File: src/components/Table/README.md [hunk 4]
```diff
@@ -237,10 +237,11 @@ const MyTable1 = withTableSettings({sortable: false})(Table);
 
 ### Options
 
-| Name     | Description                                       |      Type      | Default |
-| :------- | :------------------------------------------------ | :------------: | :-----: |
-| width    | Settings' popup width                             | `number` `fit` |         |
-| sortable | Whether or not add ability to sort settings items |   `boolean`    | `true`  |
+| Name       | Description                                         |      Type      | Default |
+| :--------- | :-------------------------------------------------- | :------------: | :-----: |
+| width      | Settings' popup width                               | `number` `fit` |         |
```
Comment (amje) [inline src/components/Table/README.md:242]:
```suggestion
| width      | Settings' popup width                               | `number` `"fit"` |         |
```

Add quotes to indicate this is a string

File: src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx [hunk 5]
```diff
@@ -283,9 +303,19 @@ export const TableColumnSetup = (props: TableColumnSetupProps) => {
...
+        filterSettings = defaultFilterSettingsFn,
     } = props;
 
     const [open, setOpen] = React.useState(false);
+    const [sortingEnabled, setSortingEnabled] = React.useState(sortable);
+    const [prevSortingEnabled, setPrevSortingEnabled] = React.useState(sortable);
+    if (sortable !== prevSortingEnabled) {
+        setPrevSortingEnabled(sortable);
+        setSortingEnabled(sortable);
+    }
```
Comment (amje) [inline src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:318]: Why changing state during render and not in effect?

**Review by amje — CHANGES_REQUESTED** [review]

**Comment by amje** [comment]
Can we remove the delay?

https://github.com/gravity-ui/uikit/assets/3368839/747c08c1-ea36-44ad-9b2f-9ff5ba2e9dac

**Comment by polikashina** [comment]
> Can we remove the delay?
> 
>  2024-06-08.22.53.39.mov

There is a delay because of [debounce](https://github.com/gravity-ui/uikit/blob/6a70f17c04f4af52e49ea5bcbe9c1f886d9c3969/src/components/useList/hooks/useListFilter.ts#L86). It is possible to set it to `0ms`

File: src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx (hunk 5 above)
Comment (polikashina) [reply src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx:318]: It should work faster than with useEffect. Component will be rerendered immediately after returning JSX before DOM updates. 
https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes

File: src/components/Table/README.md (hunk 4 above)
Comment (polikashina) [reply src/components/Table/README.md]: [done](https://github.com/gravity-ui/uikit/pull/1627/commits/e839ec965c992caa65af754928a7133cdc007b0b)

**Comment by amje** [comment]
> > Can we remove the delay?
> > 2024-06-08.22.53.39.mov
> 
> There is a delay because of [debounce](https://github.com/gravity-ui/uikit/blob/6a70f17c04f4af52e49ea5bcbe9c1f886d9c3969/src/components/useList/hooks/useListFilter.ts#L86). It is possible to set it to `0ms`

Let's set to 0. There is no reason to wait user finishes input

**Comment by polikashina** [comment]
> > > Can we remove the delay?
> > > 2024-06-08.22.53.39.mov
> > 
> > 
> > There is a delay because of [debounce](https://github.com/gravity-ui/uikit/blob/6a70f17c04f4af52e49ea5bcbe9c1f886d9c3969/src/components/useList/hooks/useListFilter.ts#L86). It is possible to set it to `0ms`
> 
> Let's set to 0. There is no reason to wait user finishes input

[done](https://github.com/gravity-ui/uikit/pull/1627/commits/6e70da7940259295db00cb58ec5b5ef373391e9c)

**Review by amje — APPROVED** [review]

**Review by Raubzeug — APPROVED** [review]

**Review by IsaevAlexandr — APPROVED** [review]

**Pull request merged**